File: benchmarks/benchmarks.R
# This file is taken from 
# primary package
library(areal)

# tidyverse packages
library(dplyr)

# spatial packages
library(sf)
library(tidycensus)
library(tigris)

# other packages
library(gridExtra)
library(microbenchmark)
library(testthat)

data(ar_stl_asthma, package = "areal")
data(ar_stl_race, package = "areal")
data(ar_stl_wards, package = "areal")
data(ar_stl_wardsClipped, package = "areal")
# Take the benchmark for the St. Louis polygons
res <- microbenchmark(
  aw_interpolate(ar_stl_wards, tid = WARD, source = ar_stl_race, sid = GEOID,
                 weight = "sum", output = "tibble", extensive = "TOTAL_E"),
  suppressWarnings(st_interpolate_aw(ar_stl_race["TOTAL_E"], ar_stl_wards, extensive = TRUE))
)
# Save the results of the benchmarking run
save(res, file = "benchmarks.RData")

# Unit: milliseconds
#                                                                                                                                          expr
#  aw_interpolate(ar_stl_wards, tid = WARD, source = ar_stl_race,      sid = GEOID, weight = "total", output = "tibble", extensive = "TOTAL_E")
#                                              suppressWarnings(st_interpolate_aw(ar_stl_race["TOTAL_E"], ar_stl_wards,      extensive = TRUE))
#       min       lq     mean   median      uq       max neval
#  68.34967 70.47232 73.24867 71.72536 74.1533 141.96041   100
#  63.22327 66.42303 68.14503 67.85898 69.4612  99.51569   100

File: benchmarks/benchmarks.jl
#=
# Benchmarks

This file contains Julia benchmarks for areal interpolation.  

## St. Louis wards
=#


# Get data from the areal package

using RData, CodecBzip2, DataFrames

function _reprocess_r_polygons!(df::DataFrame)
    geoms = map(df.geometry) do geom_array
        permutedims.(geom_array)
        GI.Polygon(GI.LinearRing.(permutedims.(geom_array) .|> vec .|> x -> reinterpret(GI.Point{false, false, Float64, Nothing}, x)))
    end
    df.geometry = geoms
    df
end

asthma = RData.load(download("https://github.com/chris-prener/areal/raw/07bda84887d9f2272babe91c2ccc1e438f27f162/data/ar_stl_asthma.rda"))["ar_stl_asthma"]
race = RData.load(download("https://github.com/chris-prener/areal/raw/07bda84887d9f2272babe91c2ccc1e438f27f162/data/ar_stl_race.rda"))["ar_stl_race"]
wards = RData.load(download("https://github.com/chris-prener/areal/raw/07bda84887d9f2272babe91c2ccc1e438f27f162/data/ar_stl_wards.rda"))["ar_stl_wards"]
wardsClipped = RData.load(download("https://github.com/chris-prener/areal/raw/07bda84887d9f2272babe91c2ccc1e438f27f162/data/ar_stl_wardsClipped.rda"))["ar_stl_wardsClipped"]
_reprocess_r_polygons!(asthma)
_reprocess_r_polygons!(race)
_reprocess_r_polygons!(wards)
_reprocess_r_polygons!(wardsClipped)

racy_wards = interpolate(Direct(), wards, race; features = (:TOTAL_E,))

# Run benchmark in Julia
using Chairmarks
using AreaInterpolation
multi_threaded_benchmark = @be interpolate(Direct(), $wards, $race; features = $((:TOTAL_E,))) seconds=3
single_threaded_benchmark = @be interpolate(Direct(), $wards, $race; features = $((:TOTAL_E,)), threaded = false) seconds=3
# Multithreaded 9ms, single-threaded 43ms, R 65ms median timings!!

# # Plotting
# Get R benchmark
r_benchmark = RData.load(joinpath(@__DIR__, "benchmarks.RData"))["res"]
r_benchmark.expr.pool.levels .= (x ->  contains(x, "st_interpolate") ? "R (sf)" : "R (areal)").(r_benchmark.expr.pool.levels)
r_benchmark
using DataFrames, Statistics
r_gdf = groupby(r_benchmark, :expr)
r_stats = [(mean(r_gdf[key].time) / 10^6, std(r_gdf[key].time) / 10^6, string(key.expr)) for key in keys(r_gdf)]

xs = [1, 2, 3, 4]
ys = [Statistics.mean(multi_threaded_benchmark).time * 10^3, Statistics.mean(single_threaded_benchmark).time * 10^3, first.(r_stats)...]
errs = [Statistics.std(multi_threaded_benchmark).time * 10^3, Statistics.std(single_threaded_benchmark).time * 10^3, getindex.(r_stats, 2)...]
xlabels = ["Multithreaded", "Single-threaded", last.(r_stats)...]

using CairoMakie, MakieThemes
with_theme(MakieThemes.bbc()) do 
    f, a, p = scatter(
        xs, ys;
        axis = (;
            title = "Benchmarking regular interpolation",
            subtitle = Makie.rich("On the St. Louis wards dataset from ", rich("areal"; font = :mono)),
            ylabel = "Median time to execute (ms)",
            xticks = (xs, xlabels),
            xticksvisible = true,
            xticklabelsvisible = true,
            ytickformat = values -> string.(round.((Int,), values)) .* " ms",
            yticks = Makie.WilkinsonTicks(7; k_min = 3, k_max = 10),
        ),
        figure = (;
            fonts = (;
                regular = "Helvetica",
                bold = "Helvetica Neue Bold",
                italic = "Helvetica Oblique",
                mono = "Fira Mono",
            ),
        )
    )
    errorbars!(
        a,
        xs,
        ys,
        errs;
        color = (:gray, 0.4)
    )
    f
end


#=

## Texas precincts

=#

